### PR TITLE
Consolidate JSON reader options

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

~Draft as it builds on https://github.com/apache/arrow-rs/pull/1451 from @sum12 .~

Closes https://github.com/apache/arrow-rs/issues/1538

# Rationale for this change
@sum12 started refactoring options into a common struct in https://github.com/apache/arrow-rs/pull/1451 and I wanted to finish that work to minimize API churn when we release the next arrow

# What changes are included in this PR?
1. Move `batch_size` to DecoderOptions as well.
2. Consolidate field handling in `ReaderBuilder`

# Are there any user-facing changes?
Yes -- if they use the low level Json decoder, the options are now encoded in a struct rather than as parameters

### CODE REVIEW

File: arrow/src/json/reader.rs [hunk 1]
```diff
@@ -1606,26 +1607,8 @@ pub struct ReaderBuilder {
...
-    /// optional HashMap of column names to format strings
-    format_strings: Option<HashMap<String, String>>,
-}
-
-impl Default for ReaderBuilder {
-    fn default() -> Self {
-        Self {
-            schema: None,
-            max_records: None,
-            batch_size: 1024,
```
Comment (alamb) [inline arrow/src/json/reader.rs:1601]: this default is moved into `DecoderOptions`

**Comment by alamb** [comment]
I would like to merge this prior to creating the arrow 12 release candidate later this week (so that we minimize the API churn) -- @nevi-me  and @houqp do you perhaps have a moment to review this?

cc @sum12

**Review by tustvold — APPROVED** [review]
It took me a while to work out why there is a separation between the `Decoder` and a `Reader` and therefore we end up with nested builders, but I think this makes sense. We want to allow people to construct a `Decoder` with arbitrary streams of `serde_json::Value` and convert them to `RecordBatch`, whereas `Reader` only supports files :+1:

**Comment by alamb** [comment]
> It took me a while to work out why there is a separation between the `Decoder` and a `Reader` and therefore we end up with nested builders, but I think this makes sense. We want to allow people to construct a `Decoder` with arbitrary streams of `serde_json::Value` and convert them to `RecordBatch`, whereas `Reader` only supports files 👍

I tried to encode this insight into the docs here: https://github.com/apache/arrow-rs/pull/1559

**Pull request merged**